src/ray/core_worker/store_provider/plasma_store_provider.cc

```diff
@@ -16,82 +16,122 @@ Status CoreWorkerPlasmaStoreProvider::Put(const RayObject &object,
+                                             const size_t data_size,
+                                             const ObjectID &object_id,
+                                             std::shared_ptr<Buffer> *data) {
+  return local_store_provider_.Create(metadata, data_size, object_id, data);
+}
+
+Status CoreWorkerPlasmaStoreProvider::Seal(const ObjectID &object_id) {
+  return local_store_provider_.Seal(object_id);
+}
+
 Status CoreWorkerPlasmaStoreProvider::Get(
     const std::vector<ObjectID> &ids, int64_t timeout_ms, const TaskID &task_id,
     std::vector<std::shared_ptr<RayObject>> *results) {
+  int64_t batch_size = RayConfig::instance().worker_fetch_request_size();
   (*results).resize(ids.size(), nullptr);
+  std::unordered_map<ObjectID, std::vector<int>> unready;
+
+  // First, attempt to fetch all of the required objects without reconstructing.
+  for (int64_t start = 0; start < ids.size(); start += batch_size) {
+    int64_t end = std::min(start + batch_size, int64_t(ids.size()));
+    const std::vector<ObjectID> ids_slice(ids.cbegin() + start, ids.cbegin() + end);
+    RAY_CHECK_OK(
+        raylet_client_->FetchOrReconstruct(ids_slice, /*fetch_only=*/true, task_id));
+    std::vector<std::shared_ptr<RayObject>> results_slice;
+    RAY_RETURN_NOT_OK(local_store_provider_.Get(ids_slice, 0, task_id, &results_slice));
+
+    // Iterate through the results from the local store, adding them to the unready
+    // map if they weren't successfully fetched from the local store (are nullptr).
+    // Keeps track of the locations of the unready object IDs in the original list
+    // (accounting for duplicates).
+    for (size_t i = 0; i < ids_slice.size(); i++) {
+      if (results_slice[i] != nullptr) {
+        (*results)[start + i] = results_slice[i];
+        continue;
+      }
+      auto it = unready.find(ids_slice[i]);
+      if (it == unready.end()) {
+        std::vector<int> v;
+        v.push_back(start + i);
+        unready.insert({ids[i], v});
+      } else {
+        it->second.push_back(start + i);
+      }
+    }
+  }
 
-  bool was_blocked = false;
-
-  std::unordered_map<ObjectID, int> unready;
-  for (size_t i = 0; i < ids.size(); i++) {
-    unready.insert({ids[i], i});
+  // If all objects were fetched already, return.
+  if (unready.empty()) {
+    return Status::OK();
   }
 
+  // If not all objects were successfully fetched, repeatedly call FetchOrReconstruct and
+  // Get from the local object store in batches. This loop will run indefinitely until the
+  // objects are all fetched if timeout is -1.
   int num_attempts = 0;
   bool should_break = false;
   int64_t remaining_timeout = timeout_ms;
-  // Repeat until we get all objects.
   while (!unready.empty() && !should_break) {
     std::vector<ObjectID> unready_ids;
     for (const auto &entry : unready) {
+      if (unready_ids.size() == batch_size) {
+        break;
+      }
```

Comment:
(commenting again in a new thread as the old one is swallowed.)
actually, I think there is another issue here. For example, if we're getting 10 batches of objects. And objects will be ready in 1s. Previously, this get operation can finish in 1s. But now, we'll need 10s (sleeping 10 times).

Considering that getting more than 10k objects is a very rare case, it's doesn't seem to be worth it to sacrifice code readability and introduce the above-mentioned problem. I'd suggest reverting this change from this PR. What do you think?

Reply:
This mirrors the existing behavior in the Python worker.

Also, I very strongly disagree that this is sacrificing code readability - the existing code is IMO much harder to understand because it conflates the semantics of the first fetch (with a 0ms timeout and `fetch_only=true`) and the subsequent repeated fetches.

Regardless, I've moved this change to a separate PR (https://github.com/ray-project/ray/pull/5342), so we can discuss it there.

Reply:
@raulchen I think batching is probably necessary. I commented on https://github.com/ray-project/ray/pull/5342.